WalletWasabi.Tests/UnitTests/StartWasabiOnSystemStartupTests.cs

```diff
@@ -1,31 +1,132 @@
+					"X-GNOME-Autostart-enabled=true");
+
 		[Fact]
 		public async Task ModifyStartupOnDifferentSystemsTestAsync()
 		{
+			UiConfig originalConfig = GetUiConfig();
+
 			if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
 			{
 				await StartupHelper.ModifyStartupSettingAsync(true);
+
+				Assert.True(RegistryKeyExists());
+
 				await StartupHelper.ModifyStartupSettingAsync(false);
+
+				Assert.False(RegistryKeyExists());
 			}
 			else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
 			{
-				await Assert.ThrowsAsync<NotImplementedException>(async () => await StartupHelper.ModifyStartupSettingAsync(true));
-				await Assert.ThrowsAsync<NotImplementedException>(async () => await StartupHelper.ModifyStartupSettingAsync(false));
+				await StartupHelper.ModifyStartupSettingAsync(true);
+
+				Assert.True(File.Exists(_filePath));
+				Assert.Equal(_expectedDesktopFileContent, GetFileContent(_filePath));
+
+				await StartupHelper.ModifyStartupSettingAsync(false);
+
+				Assert.False(File.Exists(_filePath));
 			}
 			else if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
 			{
 				await StartupHelper.ModifyStartupSettingAsync(true);
+
+				string loginItems = await GetLoginItemsAsync(_listCmd);
+
+				Assert.Contains(Constants.AppName, loginItems);
+
 				await StartupHelper.ModifyStartupSettingAsync(false);
+
+				loginItems = await GetLoginItemsAsync(_listCmd);
+
+				Assert.DoesNotContain(Constants.AppName, loginItems);
 			}
+
+			// Restore original setting for devs.
+			await StartupHelper.ModifyStartupSettingAsync(originalConfig.RunOnSystemStartup);
+		}
+
+		private UiConfig GetUiConfig()
+		{
+			string dataDir = EnvironmentHelpers.GetDataDir(Path.Combine("WalletWasabi", "Client"));
+			UiConfig uiConfig = new(Path.Combine(dataDir, "UiConfig.json"));
+			uiConfig.LoadOrCreateDefaultFile();
+
+			return uiConfig;
+		}
+
+		private bool RegistryKeyExists()
+		{
+			bool result = false;
+
+			if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+			{
+				RegistryKey? registryKey = Registry.CurrentUser.OpenSubKey(PathToRegistyKey, false);
+				result = registryKey.GetValueNames().Contains(nameof(WalletWasabi));
+			}
+
+			return result;
+		}
+
+		private string GetFileContent(string filePath)
+		{
+			return string.Join("\n", File.ReadAllLines(filePath));
+		}
+
+		private async Task<string> GetLoginItemsAsync(string cmd)
+		{
+			var escapedArgs = cmd.Replace("\"", "\\\"");
+
+			var startInfo = new ProcessStartInfo
+			{
+				FileName = "/usr/bin/env",
+				Arguments = $"sh -c \"{escapedArgs}\"",
+				RedirectStandardOutput = true,
+				UseShellExecute = false,
+				CreateNoWindow = true,
+				WindowStyle = ProcessWindowStyle.Hidden
+			};
+
+			using var process = new ProcessAsync(startInfo);
+			process.Start();
+
+			await process.WaitForExitAsync(CancellationToken.None).ConfigureAwait(false);
+
+			string output = process.StandardOutput.ReadToEnd();  // Gives back "login items Wasabi Wallet" or "login items"
```

Comment:
This line and the previous one should be switched, otherwise you risk a deadlock.

(https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.process.standardoutput?redirectedfrom=MSDN&view=net-5.0#remarks)

Reply:
Nice catch, thank you.